### PR TITLE
Profile controllers

### PR DESCRIPTION
**Pull request by rj13371**
### What this PR does (required):
added create, put and read routes as long as the proper controllers for the profile schema. 

### Screenshots / Videos (front-end only):
https://imgur.com/a/S92DbRi

### Any information needed to test this feature (required):
I tested with postman using a GET req on the route that not require a JWT or any auth and wrote the test in the test folder so you can try it out. However it is connected to a MongoDB cluster that I created 

### Any issues with the current functionality (optional):
I wasn't sure which routes would be protected with auth and what information we want to send back to the client so some props of the document may want to not be sent back.

### CODE REVIEW

File: pull_request_template.md [hunk 1]
```diff
@@ -1,11 +1,11 @@
 ### What this PR does (required):
-- list PR functionality here
+- created model for profile schema with requirements and added optional properties based on what it looks like the profile model will need in the future
```
Comment (moffatethan) [inline pull_request_template.md]: Just so you are aware, the `pull_request_template.md` will be what is prefilled in the comment when you create a new PR on GitHub. You don't need to edit this file directly for every pull request.

File: server/controllers/profile.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+// - Given an ID and new parameters, update the profile
+
+// - GET
+
+// - Given an ID, return profile with that ID
+
+// - GET all
+
+// - A list of profiles
```
Comment (moffatethan) [inline server/controllers/profile.js]: I like to see you planning your work beforehand, but after it's completed do we need these comments here still?

File: server/controllers/profile.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+const Profile = require("../models/Profile");
+const asyncHandler = require("express-async-handler");
+
+
+exports.searchProfiles = asyncHandler(async (req, res, next) => {
+
+  const profiles = await Profile.find({});
+
+  res.status(200).json({ profiles: profiles });
+});
```
Comment (moffatethan) [inline server/controllers/profile.js]: Search denotes the ability to pass in data and return potential results. Is the naming incorrect here 🤔. Or do we want to pass down some simple things such as the username or names. 

If this is just simply for getting all the profiles on the database then perhaps a new name would be more clear.

File: server/controllers/profile.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+// - A list of profiles
+const Profile = require("../models/Profile");
+const asyncHandler = require("express-async-handler");
+
+
+exports.searchProfiles = asyncHandler(async (req, res, next) => {
+
+  const profiles = await Profile.find({});
+
+  res.status(200).json({ profiles: profiles });
```
Comment (moffatethan) [inline server/controllers/profile.js]: nit: this can be `res.status(200).json({ profiles });`

Comment (moffatethan) [reply server/controllers/profile.js]: ditto on other parts of the code that look similar.

File: server/controllers/profile.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  res.status(200).json({ profile: profile });
+});
+
+exports.updateProfile = asyncHandler(async (req, res, next) => {
+
+  try {
+
+    const profile = await Profile.findByIdUpdate(req.params.id, { ...req.body.profile }) //update props based on what was sent in the req body
```
Comment (moffatethan) [inline server/controllers/profile.js]: Shouldn't this be `findByIdAndUpdate`?

File: server/controllers/profile.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  try {
+
+    const profile = new Profile(req.body.profile)
+
+    await profile.save();
+
+    res.status(200).json({ profile: profile });
+  } catch (e) {
+    res.status(404).json({ message: e });
+  }
```
Comment (moffatethan) [inline server/controllers/profile.js]: Confused on `404` being thrown here. If we're creating a profile how would we ever get not found? Is this related to the user reference?

File: server/models/Profile.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
+const mongoose = require('mongoose');
+const Schema = mongoose.Schema;
+
```
Comment (moffatethan) [inline server/models/Profile.js]: This code is repeated from the profile model branch. Let's base this branch. Read below 👍🏻 

[How to change the base branch of a PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request)

File: server/routes/profile.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+const protect = require("../middleware/auth");
+const { searchProfiles,searchProfileById, updateProfile, createProfile } = require("../controllers/profile");
+
+router.route("/profiles").get(searchProfiles);
+
+router.route("/create_profile").post(protect, createProfile);
+
+router.route("/:id").get(searchProfileById); // im not sure if we are using this route to display profiles to a user once they are logged
+//in and see their own profile, or just to see other users profiles, its not clear to me, so if we want to display
+//all props of the profile to any user when they search, then we obviously do not want to send sensitive details of the persons profile
```
Comment (moffatethan) [inline server/routes/profile.js]: We actually wouldn't necessarily be searching by the id directly but rather their name, location, etc. Of course, we don't have any advanced searching like Elasticsearch so our implementation will be simplified.

Comment (moffatethan) [reply server/routes/profile.js]: But this route is still needed for fetching profile information on say our messages.

File: server/routes/profile.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+//we can add protect middleware if we only want to show profile details to the logged in user, but for now there is no protect auth
```
Comment (moffatethan) [inline server/routes/profile.js]: Yeah, we'll want to protect this resource.

**Review by moffatethan — CHANGES_REQUESTED** [review]
Good job so far, some comments to address and please base the branch so we don't see repeated files in the git diff on the PR. Thanks Roland!

File: server/models/Profile.js (hunk 7 above)
Comment (rj13371) [reply server/models/Profile.js]: I changed the base to profile model branch and resolved the changes, is this correct?

File: server/controllers/profile.js (hunk 6 above)
Comment (rj13371) [reply server/controllers/profile.js]: I changed the status code to 400 but im not sure what status code should be thrown for a error we don't know we will encounter. What do you suggest?

File: server/routes/profile.js (hunk 8 above)
Comment (rj13371) [reply server/routes/profile.js]: So if we are going to make a Message schema with a reference to profile, I should keep this route as it is correct? We would probably search the profile by ID that is referenced in the Message document.

File: server/controllers/profile.js (hunk 6 above)
Comment (moffatethan) [reply server/controllers/profile.js]: `400` should be fine, my only thought is the validation failing before saving happening. So we'll keep this is. Ideally, we would implement middleware validation for better integrity of our incoming data expectations in the API.

File: server/routes/profile.js (hunk 8 above)
Comment (moffatethan) [reply server/routes/profile.js]: Sure, this should be fine. We'll finalize the implementation in planning phases for the message feature. 👍🏻

**Review by moffatethan — APPROVED** [review]
Good work!

File: server/test/ping.spec.js [hunk 10]
```diff
@@ -1,22 +1,15 @@
...
-      .end((err, res) => {
-        res.should.have.status(200);
-        res.body.should.have
-          .property("response")
-          .eql("Server is running. Message received: hello");
-        done();
-      });
+describe("profile GET route", function() {
+
+  var url = "http://localhost:3001/profiles/profiles"
```
Comment (bonnieli) [inline server/test/ping.spec.js:7]: we probably wanna avoid using `var` 

this looks like it is also dependant on a _running_ server which we would like to avoid - tests should be able to run independently
